File: backend/app/layouts.py
"""
Quais layouts existem, e qual rubrica pode usar cada um.

Dois deles são feitos para um tipo de conteúdo e não fazem sentido no outro:

  - **cover** monta capa + banner sobre o vídeo, e a capa é escolhida pelo
    quadro mais expressivo de um ROSTO falando. Num gameplay ela pega um frame
    de tela de jogo, que não diz nada.
  - **streamer** empilha facecam sobre gameplay. Num podcast não há facecam
    separada para empilhar — o vídeo inteiro já é a câmera.

Os outros dois servem a qualquer conteúdo porque não presumem nada sobre ele:
recortar no centro e não recortar nada.

Esta lista é espelhada em `frontend/src/lib/layouts.ts`. Divergirem é bug, e
`tests/test_layouts.py` compara as duas.
"""

from typing import Annotated

from pydantic import AfterValidator

#: layout → rubricas que o aceitam. Vazio = serve a todas.
_LAYOUTS: dict[str, tuple[str, ...]] = {
    # Capa (rosto expressivo) + banner de título + vídeo com face tracking.
    "cover": ("podcast",),
    # Facecam empilhada sobre o gameplay, com faixa no meio.
    "streamer": ("gameplay", "siege"),
    # Recorte 9:16 centralizado, sem camada nenhuma.
    "crop": (),
    # Sem recorte: o enquadramento e a resolução da fonte, só cortado no tempo.
    "original": (),
}

ALL_LAYOUTS: tuple[str, ...] = tuple(_LAYOUTS)

#: O que cada modo faz, na linguagem de quem escolhe.
LAYOUT_LABELS: dict[str, tuple[str, str]] = {
    "cover": ("Capa + banner", "Capa com o rosto, título em destaque e o vídeo embaixo"),
    "streamer": ("Facecam + gameplay", "A câmera em cima, o jogo embaixo"),
    "crop": ("Crop vertical", "Recorte 9:16 no centro, sem camadas"),
    "original": (
        "Layout original",
        "Sem reenquadrar: o corte sai como está no vídeo de origem",
    ),
}


def layouts_for(source_type: str | None) -> tuple[str, ...]:
    """Layouts que esta rubrica aceita, na ordem de exibição."""
    nicho = (source_type or "").lower()
    return tuple(
        layout
        for layout, rubricas in _LAYOUTS.items()
        if not rubricas or nicho in rubricas
    )


def layout_allowed(layout: str | None, source_type: str | None) -> bool:
    return (layout or "") in layouts_for(source_type)


def _ensure_known_layout(value: str) -> str:
    if value not in _LAYOUTS:
        raise ValueError(
            f"Layout inválido: escolha um de {', '.join(ALL_LAYOUTS)}"
        )
    return value


#: Tipo para o campo de layout. Só valida que o modo EXISTE; a combinação com a
#: rubrica depende de dois campos e é conferida em quem os tem à mão.
LayoutField = Annotated[str, AfterValidator(_ensure_known_layout)]

File: frontend/src/lib/layouts.ts
import { PERSONAL_LAYOUT_RUBRICS } from "@/personal";
import type { LayoutMode, SourceType } from "./types";

/**
 * Quais layouts existem, e qual rubrica pode usar cada um.
 *
 * ESPELHO de `backend/app/layouts.py`. As duas cópias existem porque a decisão
 * é tomada nos dois lados — a tela precisa saber o que oferecer, o servidor
 * precisa recusar o que não serve. Divergirem é bug: um layout liberado só aqui
 * vira um botão que o servidor recusa; só lá, uma opção que ninguém vê.
 * `backend/tests/test_layouts.py` compara os dois.
 *
 * Dois modos presumem coisas sobre o conteúdo: a **capa** é escolhida pelo
 * quadro mais expressivo de um rosto falando, e a **facecam empilhada** precisa
 * de uma câmera separada do gameplay. Os outros dois não presumem nada.
 */

/**
 * layout → rubricas PÚBLICAS que o aceitam. Lista vazia = serve a todas.
 *
 * Só o que existe nos dois builds. As rubricas pessoais entram por
 * `PERSONAL_LAYOUT_RUBRICS`, que é vazia no stub — senão o nome de uma feature
 * que o build público não tem viajaria no bundle dele como valor desta tabela.
 */
export const BASE_LAYOUT_RUBRICS: Record<LayoutMode, SourceType[]> = {
  cover: ["podcast"],
  streamer: ["gameplay"],
  crop: [],
  original: [],
};

/** A tabela completa deste build: a pública mais o que o pessoal acrescenta. */
export const LAYOUT_RUBRICS: Record<LayoutMode, SourceType[]> = Object.fromEntries(
  (Object.keys(BASE_LAYOUT_RUBRICS) as LayoutMode[]).map((layout) => [
    layout,
    // Lista vazia é "serve a todas": acrescentar uma rubrica a ela seria
    // RESTRINGIR o layout, e não ampliá-lo.
    BASE_LAYOUT_RUBRICS[layout].length === 0
      ? []
      : [...BASE_LAYOUT_RUBRICS[layout], ...(PERSONAL_LAYOUT_RUBRICS[layout] ?? [])],
  ]),
) as Record<LayoutMode, SourceType[]>;

export const LAYOUT_LABELS: Record<LayoutMode, { nome: string; descricao: string }> = {
  cover: {
    nome: "Capa + banner",
    descricao: "Capa com o rosto, título em destaque e o vídeo embaixo",
  },
  streamer: {
    nome: "Facecam + gameplay",
    descricao: "A câmera em cima, o jogo embaixo",
  },
  crop: {
    nome: "Crop vertical",
    descricao: "Recorte 9:16 no centro, sem camadas",
  },
  original: {
    nome: "Layout original",
    descricao: "Sem reenquadrar: o corte sai como está no vídeo de origem",
  },
};

/** Os layouts que esta rubrica aceita, na ordem de exibição. */
export function layoutsFor(source: SourceType): LayoutMode[] {
  return (Object.keys(LAYOUT_RUBRICS) as LayoutMode[]).filter((layout) => {
    const rubricas = LAYOUT_RUBRICS[layout];
    return rubricas.length === 0 || rubricas.includes(source);
  });
}

export function layoutAllowed(layout: LayoutMode, source: SourceType): boolean {
  return layoutsFor(source).includes(layout);
}
